onnxruntime/core/graph/graph.cc

```diff
@@ -2100,6 +2090,26 @@ Status Graph::VerifyNodeAndOpMatch(const ResolveOptions& options) {
   return Status::OK();
 }
 
+void Graph::InitFunctionBodyForNode(Node& node) {
+  if (node.op_ && (node.op_->HasFunction() || node.op_->HasContextDependentFunction())) {
+    onnx::FunctionProto onnx_function_proto;    
+    if (node.op_->HasContextDependentFunction()) {
+      NodeProto node_proto;
+      node.ToProto(node_proto);
+      onnx::FunctionBodyBuildContextImpl function_body_ctx(node_proto);
+      node.op_->BuildContextDependentFunction(function_body_ctx, onnx_function_proto);
+    } else {
+      onnx_function_proto = *(node.op_->GetFunction());
+    }
+
+    auto func_ptr = onnxruntime::make_unique<onnxruntime::FunctionImpl>(*this, node.Index(), onnx_function_proto,
+                                                                        logger_);
+
+    function_container_.emplace_back(std::move(func_ptr));
+    node.SetFunctionBody(*function_container_.back());
+  }
+}
+
```

Comment:
nit: all of this logic seems to be at the node level with the exception of saving the func_ptr. would it make more sense for Node to have a unique_ptr<FunctionImpl> for the function body so that all this can move into Node::GetFunctionBody? or is the FunctionImpl potentially shared by other nodes?

Reply:
Per our offline conversation - There is one more issue regarding functions in ort. Function is used in 2 different contexts - 1. ONNX Function 2. sub graphs of fused nodes created for execution providers based on EPs capability. Node Type Fused is only used for 1. where as ONNX functions node type is set as primitive... this is confusing because in ONNX, functions are defined as a collection of primitive ops. Long story short I am planning to fix this as part of a different PR and will address your comment in the same PR.